### ISSUE TITLE
VALUES() must cast NULL literals in the first row, if the type is known and the RDBMS cannot infer it

### ISSUE THREAD

**Issue by baev** [jOOQ/jOOQ#11015]
The issue is that when query has bind variable of type bigint with all null values, specified type from `val(someNullValue, SQLDataType.BIGINT)` is ignored during bind variables inlining.

### Expected behavior

The query works as excepted and explicit casts are added during bind variables inlining.

### Actual behavior

No explicit cast causes a bad grammar exception.

### Steps to reproduce the problem

The show case:

```sql
create table a(id bigserial primary key, name text);
create table b(id bigserial primary key, a_id bigint references a(id), name text);

insert into a(name) values ('first'), ('second');
insert into b(a_id, name) select c1, c2 from (values (null, 'b1'), (null, 'b2')) c(c1, c2);
```

PostgreSQL can't determinate the type of `c1` as all provided values are nulls. Having at least one non-null value fixes the query (Or explicit cast `null::bigint`);

So in JOOQ we are using `val` with specified type for nullable columns, and it works just fine except the case when JOOQ inline all bind variables because of bind variables limit.

```java
final Row2<Long, String>[] rows = requests.stream()
                .map(rq -> row(
                        val(rq.geA(), SQLDataType.BIGINT),
                        val(rq.getB()))
                ).toArray(Row2[]::new);

 final Table<Record2<Long, String>> data = values(rows);
 final Record2<Long, String> dataRec = data.newRecord();

dsl.insertInto(SOME)
    .columns(SOME.C_BIGINT, SOME.C_TEXT)
    .select(dataRec.field1(), dataRec.field2())
    .from(data)
    .execute();
```

### Versions

- jOOQ: 3.14.0
- Java: 11.0.2
- Database (include vendor): PostgreSQL
- OS: MacOS
- JDBC Driver (include name if inofficial driver):

**Comment by baev** [comment]
the workaround is to add explicit casts by yourself:

```java
Objects.isNull(rq.getA())
                                ? castNull(SQLDataType.BIGINT)
                                : val(rq.getA(), SQLDataType.BIGINT)
```java

**Comment by lukaseder** [comment]
Thanks for your report. That's a curious issue, and I think the first time we ran into having to cast things when inlining, in PostgreSQL - although it does make me think if there are other issues where `null` doesn't work and should be replaced by `cast(null as <type>)`. For example, this also doesn't work:

```sql
select 1 where 1 is distinct from (select null)
```

The error being:

> SQL Error [42883]: ERROR: operator does not exist: integer = text
>  Hint: No operator matches the given name and argument types. You might need to add explicit type casts.
>  Position: 18

**Comment by lukaseder** [comment]
There are more edge cases as this isn't strictly related to the literal being the `null` literal (though it probably happens more frequently with that). For example, this fails, too:

```sql
select 1 where array['a']::varchar(1)[] is distinct from (select array['a'])
```

With:

> SQL Error [42883]: ERROR: operator does not exist: character varying[] = text[]
>  Hint: No operator matches the given name and argument types. You might need to add explicit type casts.
>  Position: 41

This is a frequent error otherwise in jOOQ generated SQL, because we tend to create `varchar[]` types instead of the more frequently used `text[]` types in PostgreSQL, even when casting bind parameters.

**Comment by lukaseder** [comment]
I'm fixing a related issue where not all values in a row are `null`: https://github.com/jOOQ/jOOQ/issues/5851. As soon as one value is not null, and if the dialect still can't infer it, then we'll cast.

Once that's done, I'll look at the `null` case separately. While the other issues I've mentioned deserve some attention, too, I prefer looking into them separately, not in the context of this `VALUES()` specific behaviour: https://github.com/jOOQ/jOOQ/issues/11514

**Comment by lukaseder** [comment]
Related: https://github.com/jOOQ/jOOQ/issues/14199

**Comment by lukaseder** [comment]
So, the logic implemented here will be: When the first row renders `null` literals and data types are available on the `VALUES` expression (from any row), then we cast that `null` literal explicitly. Explicitly cast `null` literals aren't re-cast, assuming users already implemented the logic themselves, or prefer a different type than the one inferred.

**Comment by lukaseder** [comment]
I'll also attempt to fix the case where types are not specified explicitly on the bind values:

```java
row(
  val(rq.geA()),
  val(rq.getB())
)
```

In that case, the compiler can infer the type, but the `VALUES` clause itself doesn't know its own row type at runtime. But since we're inserting this `VALUES` into a known table with explictly specified `columns()`, we can still coerce types, possibly with an explicit `CAST` in the `SELECT` clause.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0. I'll backport the fix early next week.

**Comment by lukaseder** [comment]
Fixed in 3.17.9 (https://github.com/jOOQ/jOOQ/issues/14758), and 3.16.15 (https://github.com/jOOQ/jOOQ/issues/14759)
